### PR TITLE
feature: add the Kms type to provide a set of key mgmt functions

### PR DESCRIPTION
**Pull request by jimlambrt**

### CODE REVIEW

File: extras/kms/repository.go [hunk 1]
```diff
@@ -17,7 +17,8 @@ const (
 	// DefaultLimit is the default for results
 	DefaultLimit = 10000
 
-	DefaultWrapperSecret = "secret"
+	// DefaultWrapperSecret defines a default secret for testing
+	DefaultWrapperSecret = "secret1234567890"
```
Comment (johanbrandhorst) [inline extras/kms/repository.go:21]: Do we need to define this in this package, vs putting it in the test file?

File: extras/kms/key_purpose.go [hunk 2]
```diff
@@ -18,3 +22,29 @@ func reservedKeyPurpose() []string {
...
+		purposesMap[purpose] = struct{}{}
+	}
+	purposes = make([]KeyPurpose, 0, len(purposesMap))
+	for purpose := range purposesMap {
+		purposes = append(purposes, purpose)
+	}
+	return purposes
+}
+
+func purposeListContains(haystack []KeyPurpose, needle KeyPurpose) bool {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply extras/kms/key_purpose.go:43]: not yet.... a future refactor yes once 1.18+ is widely used.

File: extras/kms/repository.go (hunk 1 above)
Comment (jimlambrt) [reply extras/kms/repository.go:21]: it's used in both internal and external pkg tests.

File: extras/kms/data_key.go [hunk 3]
```diff
@@ -8,6 +8,8 @@ import (
 	"github.com/hashicorp/go-dbw"
 )
 
+//  DataKey represents the DEKs (keys to encrypt data) of the system and must
+//  have a parent root key and a purpose.
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply extras/kms/data_key.go:12]: not sure how that happen.. but it makes me chuckle that you noticed it.

File: extras/kms/kms.go [hunk 4]
```diff
@@ -1,13 +1,348 @@
...
+	const op = "kms.New"
+	if repo == nil {
+		return nil, fmt.Errorf("%s: missing underlying repo: %w", op, ErrInvalidParameter)
+	}
+	purposes = append(purposes, KeyPurposeRootKey)
+	RemoveDuplicatePurposes(purposes)
+
+	return &Kms{
+		purposes: purposes,
+		repo:     repo}, nil
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply extras/kms/kms.go:56]: happy to change it.

File: extras/kms/key_purpose.go [hunk 5]
```diff
@@ -18,3 +22,29 @@ func reservedKeyPurpose() []string {
 		string(KeyPurposeRootKey),
 	}
 }
+
+// RemoveDuplicatePurposes will de-dup a set of key purposes
+func RemoveDuplicatePurposes(purposes []KeyPurpose) []KeyPurpose {
+	purposesMap := make(map[KeyPurpose]struct{}, len(purposes))
+	for _, purpose := range purposes {
+		purpose = KeyPurpose(strings.TrimSpace(string(purpose)))
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply extras/kms/key_purpose.go:30]: moved to purpose.trimSpace()

ty!

**Review by johanbrandhorst — APPROVED** [review]

File: extras/kms/kms.go [hunk 6]
```diff
@@ -1,5 +1,273 @@
...
+//
+// TODO: If we support more than one, e.g. for encrypting against many in case
+// of a key loss, there will need to be some refactoring here to have the values
+// being stored in the struct be a multiwrapper, but that's for a later project.
+func (k *Kms) AddExternalWrapper(ctx context.Context, purpose KeyPurpose, wrapper wrapping.Wrapper) error {
+	return k.addKey(ctx, ExternalWrapperCache, purpose, wrapper)
+}
+
+// returns ErrKeyNotFound when a key for the given purpose is not found.
+func (k *Kms) GetExternalWrapper(ctx context.Context, purpose KeyPurpose) (wrapping.Wrapper, error) {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply extras/kms/kms.go:115]: ty!

File: extras/kms/kms.go [hunk 7]
```diff
@@ -1,5 +1,273 @@
...
+// of the kms, e.g. in a configuration file.
+//
+// TODO: If we support more than one, e.g. for encrypting against many in case
+// of a key loss, there will need to be some refactoring here to have the values
+// being stored in the struct be a multiwrapper, but that's for a later project.
+func (k *Kms) AddExternalWrapper(ctx context.Context, purpose KeyPurpose, wrapper wrapping.Wrapper) error {
+	return k.addKey(ctx, ExternalWrapperCache, purpose, wrapper)
+}
+
+// returns ErrKeyNotFound when a key for the given purpose is not found.
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply extras/kms/kms.go:114]: ty!

**Comment by jimlambrt** [comment]
> if you plan on keeping these individual commits, this should be moved to the previous commit where this function was originally added, if you plan on squashing these you can just ignore this. 

ty for catching this.  the repo requires a squash merge..

**Review by tmessi — APPROVED** [review]

**Pull request merged**